lib/fhir/common.js

```diff
@@ -119,6 +110,155 @@ module.exports = exports = (mongo) => {
     throw new Error(`Unknown ID generator '${generator}'. Valid options are 'uuidv1' or 'uuidv4'`)
   }
 
+  function resolveReference (reference, callback) {
+    const matches = REFERENCE_REGEX.exec(reference)
+    if (!matches) {
+      return callback(new Error(`Invalid resource reference "${reference}"`))
+    }
+    mongo.getDB((err, db) => {
+      if (err) {
+        return callback(err)
+      }
+      const resourceType = matches[4]
+      const resourceId = matches[5]
+      const versionId = matches[7]
+      if (versionId) {
+        db
+          .collection(`${resourceType}_history`)
+          .findOne({ id: resourceId, 'meta.versionId': versionId }, callback)
+      } else {
+        db.collection(resourceType).findOne({ id: resourceId }, callback)
+      }
+    })
+  }
+
+  function resolveNestedProperties (resultList, includeReference, callback) {
+    const promises = []
+
+    const evaluateProperties = (property, nextProperty) => {
+      if (property == null) {
+        return callback(new Error('Undefined resource property.'))
+      }
+
+      _.castArray(property).map((prop) => {
+        if (nextProperty[0] && prop[nextProperty[0]] != null) {
+          evaluateProperties(prop[nextProperty[0]], nextProperty.slice(1))
+        } else if (prop.reference) {
+          promises.push(new Promise((resolve, reject) => {
+            resolveReference(prop.reference, (err, data) => {
+              if (err) {
+                reject(err)
+              } else {
+                resolve(data)
+              }
+            })
+          }))
+        } else {
+          return callback(new Error(`Invalid resource reference: "${JSON.stringify(prop)}"`))
```

Comment:
This isn't going to work. The return here will only exit from the current iteration of map. That means that execution of the rest of the elements will continue and the callback will eventually be called multiple times. We need to be smarter about handling this error.